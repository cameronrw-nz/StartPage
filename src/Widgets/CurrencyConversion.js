import React, { useState, useEffect } from "react";

import * as Api from "../Api/Api";

const availableCurrencies = ["NZD", "THB", "USD"];

function CurrencyConversion() {
    const [fromCurrency, setFromCurrency] = useState("NZD");
    const [toCurrency, setToCurrency] = useState("THB");
    const [fromAmount, setFromAmount] = useState(1);
    const [toAmount, setToAmount] = useState(1);
    const [fromTo, setFromTo] = useState(1);

    useEffect(() => {
        getExchangeRate();
    }, []);

    function getExchangeRate(isSettingTo = true) {
        Api.getCurrencyConversion(fromCurrency, toCurrency).then(response => {
            if (
                !response ||
                !response.data ||
                !response.data[`${fromCurrency}_${toCurrency}`]
            ) {
                return;
            }
            setFromTo(response.data[`${fromCurrency}_${toCurrency}`].val);
            if (isSettingTo) {
                setToAmount(
                    (
                        parseFloat(fromAmount) *
                        response.data[`${fromCurrency}_${toCurrency}`].val
                    ).toFixed(2)
                );
            } else {
                setFromAmount(
                    (
                        parseFloat(toAmount) /
                        response.data[`${fromCurrency}_${toCurrency}`].val
                    ).toFixed(2)
                );
            }
        });
    }

    function onToChanged(e) {
        setToAmount(e.target.value);
        setFromAmount((parseFloat(e.target.value) * fromTo).toFixed(2));
    }

    function onFromChanged(e) {
        setToAmount((parseFloat(e.target.value) / fromTo).toFixed(2));
        setFromAmount(e.target.value);
    }

    function onFromCurrencyChanged() {
        var currencyIndex = availableCurrencies.indexOf(fromCurrency);
        var newIndex = (currencyIndex + 1) % availableCurrencies.length;
        setFromCurrency(availableCurrencies[newIndex]);
        getExchangeRate();
    }

    function onToCurrencyChanged() {
        var currencyIndex = availableCurrencies.indexOf(toCurrency);
        var newIndex = (currencyIndex + 1) % availableCurrencies.length;
        setToCurrency(availableCurrencies[newIndex]);
        getExchangeRate(false);
    }

    return (
        <div className="sidepanel-content">
            <h2>Currency Converter</h2>
            <div className="sidepanel-link-input">
                <input
                    type="text"
                    placeholder="THB"
                    value={fromAmount}
                    onInput={onFromChanged}
                />
                <div
                    className="currency"
                    style={{ marginRight: "4px" }}
                    onClick={onFromCurrencyChanged}
                >
                    {fromCurrency}
                </div>
                <input
                    type="text"
                    placeholder="NZD"
                    value={toAmount}
                    onInput={onToChanged}
                />
                <div className="currency" onClick={onToCurrencyChanged}>
                    {toCurrency}
                </div>
            </div>
        </div>
    );
}

export default CurrencyConversion;
