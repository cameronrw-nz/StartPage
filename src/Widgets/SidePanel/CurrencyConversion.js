import React, { useState, useEffect } from "react";

import * as Api from "../../Api/Api"

function CurrencyConversion() {
    const [thb, setThb] = useState(1)
    const [nzd, setNzd] = useState(1)
    const [nzdToThb, setNzdToThb] = useState(1)

    useEffect(() => {
        Api.getCurrencyConversion()
            .then(
                response => {
                    if (!response || !response.data || !response.data.NZD_THB) {
                        return;
                    }
                    setNzdToThb(response.data.NZD_THB.val);
                    setThb((parseFloat(nzd) * response.data.NZD_THB.val).toFixed(2));
                });
    });

    function onNZDChanged(e) {
        setNzd(e.target.value);
        setThb((parseFloat(e.target.value) * nzdToThb).toFixed(2));
    }

    function onTHBChanged(e) {
        setNzd((parseFloat(e.target.value) / nzdToThb).toFixed(2));
        setThb(e.target.value);
    }
    console.log(`NZD: ${nzd}, THB: ${thb}`);
    return (
        <div className="sidepanel-content">
            <h2>Currency Converter</h2>
            <input type="text" placeholder="THB" name="thb" value={thb} onInput={onTHBChanged} />
            <input type="text" placeholder="NZD" name="nzd" value={nzd} onInput={onNZDChanged} />
        </div>
    );
}

export default CurrencyConversion;