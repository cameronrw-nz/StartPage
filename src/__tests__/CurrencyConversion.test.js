jest.mock('axios');

import axios from 'axios';
import React from 'react';
import {render, act} from 'react-testing-library';

import CurrencyConversion from "../Widgets/SidePanel/CurrencyConversion";

describe("Currency Component Render", () => {
    it("shows a list of users", async () => {
        const users = {"NZD_THB":{"val":21.45094}};
        const resp = {data: users};
        axios.get.mockResolvedValue(resp);
        expect(axios.get).toHaveBeenCalledTimes(0);
    
        let component;
        
        await act(async () => {
            component =await  render(<CurrencyConversion />);

        });
    
        expect(axios.get).toHaveBeenCalledTimes(1);
        expect(component.container.firstChild).toMatchSnapshot();
    });
  });
