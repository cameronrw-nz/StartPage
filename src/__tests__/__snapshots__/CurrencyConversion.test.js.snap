// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Currency Component Render shows a list of users 1`] = `
<div
  class="sidepanel-content"
>
  <h2>
    Currency Converter
  </h2>
  <input
    name="thb"
    placeholder="THB"
    type="text"
    value="1"
  />
  <input
    name="nzd"
    placeholder="NZD"
    type="text"
    value="1"
  />
  <div>
    1
  </div>
  <div>
    1
  </div>
</div>
`;
